jwst/outlier_detection/utils.py

```diff
@@ -20,177 +26,365 @@
 _ONE_MB = 1 << 20
 
 
+def nanmedian3D(cube, overwrite_input=True):
+    """Compute the median of a cube ignoring warnings and with memory efficiency.
+    np.nanmedian always uses at least 64-bit precision internally, and this is too
+    memory-intensive. Instead, loop over the median calculation to avoid the
+    memory usage of the internal upcasting and temporary array allocation.
+    The additional runtime of this loop is indistinguishable from zero,
+    but this loop cuts overall step memory usage roughly in half for at least one
+    test association.
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings(action="ignore",
+                                message="All-NaN slice encountered",
+                                category=RuntimeWarning)
+        output_arr = np.empty(cube.shape[1:], dtype=np.float32)
+        for i in range(output_arr.shape[0]):
+            # this for loop looks silly, but see docstring above
+            output_arr[i,:] = np.nanmedian(cube[:,i,:], axis=0, overwrite_input=overwrite_input).astype(np.float32)
```

Comment:
Is this cast needed? If I run the following I see a float32:
```
import numpy as np
arr = np.ones((5, 30), dtype='f4')
result = np.nanmedian(arr, axis=0, overwrite_input=True)
assert result.dtype == np.dtype('f4')
```
Also `nanmedian` supports an `out` argument so I think this could be:
```suggestion
            np.nanmedian(cube[:,i,:], axis=0, overwrite_input=overwrite_input, out=output_arr[i,:])
```

Reply:
Changed